google-api-client/src/main/java/com/google/api/client/googleapis/mtls/MtlsUtils.java

```diff
@@ -0,0 +1,162 @@
+      String useClientCertificate = envProvider.getenv(GOOGLE_API_USE_CLIENT_CERTIFICATE);
+      return "true".equals(useClientCertificate);
+    }
+
+    @Override
+    public String getKeyStorePassword() {
+      return "";
+    }
+
+    @Override
+    public KeyStore getKeyStore() throws IOException, GeneralSecurityException {
+      // Load the cert provider command from the json file.
+      InputStream stream;
+      try {
+        stream = new FileInputStream(metadataPath);
+      } catch (FileNotFoundException ignored) {
+        // file doesn't exist
+        return null;
+      }
+
+      List<String> command = extractCertificateProviderCommand(stream);
+
+      // Call the command.
+      Process process = new ProcessBuilder(command).start();
+      int exitCode = -1;
+      try {
+        // Run the command and timeout after 1000 milliseconds.
+        exitCode = runCertificateProviderCommand(process, 1000);
+      } catch (InterruptedException e) {
+        throw new IOException("Interrupted executing certificate provider command", e);
+      }
+      if (exitCode != 0) {
+        throw new IOException(
+            String.format("Failed to execute cert provider command with exit code: %d", exitCode));
+      }
+
+      // Parse input certificates from shell command
+      return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+    }
+
+    @VisibleForTesting
+    static List<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+        throws IOException {
+      JsonParser parser = Utils.getDefaultJsonFactory().createJsonParser(contextAwareMetadata);
+      ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+      return json.getCommands();
+    }
+
+    @VisibleForTesting
+    static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+        throws IOException, InterruptedException {
+      long startTime = System.currentTimeMillis();
+      long remainTime = timeoutMilliseconds;
+      boolean terminated = false;
+
+      do {
+        try {
+          // Check if process is terminated by pooling the exitValue, which throws
+          // IllegalThreadStateException if not terminated.
+          commandProcess.exitValue();
+          terminated = true;
+          break;
+        } catch (IllegalThreadStateException ex) {
+          if (remainTime > 0) {
+            Thread.sleep(Math.min(remainTime + 1, 100));
+          }
+        }
+        remainTime = remainTime - (System.currentTimeMillis() - startTime);
+      } while (remainTime > 0);
+
+      if (!terminated) {
+        commandProcess.destroy();
+        throw new IOException("cert provider command timed out");
+      }
+
+      return commandProcess.exitValue();
+    }
+  }
+
+  private static final MtlsProvider MTLS_PROVIDER = new DefaultMtlsProvider();
+
+  /**
+   * Returns the default MtlsProvider instance.
+   *
+   * @return The default MtlsProvider instance.
```

Comment:
nit: "the default MtlsProvider instance" per javadoc guidelines
no period

Reply:
done